Limit CVPixelBuffer instances to reasonable sizes

mwyrzykowski opened the pull request:
#### 673419689ef0d42c50b21334edf62fd179702f16
<pre>
Limit CVPixelBuffer instances to reasonable sizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=297470">https://bugs.webkit.org/show_bug.cgi?id=297470</a>
<a href="https://rdar.apple.com/158231386">rdar://158231386</a>

Reviewed by Youenn Fablet and Jer Noble.

Limit CVPixelBuffer instances to reasonable sizes.

16k x 16k is the IOSurface maximum size and considerably
smaller than 8k video.

* Source/WebCore/platform/graphics/cv/CVUtilities.mm:
(WebCore::createBlackPixelBuffer):

Canonical link: <a href="https://commits.webkit.org/301112@main">https://commits.webkit.org/301112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1a23e4fe670e78cb6e809ec711f4e038f2ef377

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124947 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35353 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131797 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76828 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53183 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95095 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127901 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36153 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111741 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75643 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35088 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29894 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75277 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105916 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30125 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134470 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51775 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39576 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103576 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103352 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26317 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48684 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26975 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48796 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51655 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57456 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51036 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54391 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52729 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/cv/CVUtilities.mm:
```diff
@@ -180,6 +180,9 @@ void setOwnershipIdentityForCVPixelBuffer(CVPixelBufferRef pixelBuffer, const Pr
 
     NSDictionary *pixelAttributes = @{ (__bridge NSString *)kCVPixelBufferIOSurfacePropertiesKey : @{ } };
 
+    if (width * height > 35651584)
```
@youennf should I have the limit here or in `LibWebRTCCodecs::encodeFrame`?

mwyrzykowski replied:
@jhertz-apple reported another repro which occurs in `SharedVideoFrameReader::readBuffer` (which proceeds to call `WebCore::createBlackPixelBuffer`) so perhaps leaving the check in `WebCore::createBlackPixelBuffer` is most appropriate

jernoble reviewed (approved):
> 35,651,584 appears to be the width * height limit on AV1.

This might be the case for AV1, but not AV2, or VVC, or some other future codec; and that's the problem with hard coded limits like this. That said, it does seem reasonable to limit the size of these texture, but we should continue to work with the platform to identify and resolve the underlying problem so that workarounds like this aren't necessary.

youennf commented on Source/WebCore/platform/graphics/cv/CVUtilities.mm:
```diff
@@ -180,6 +180,11 @@ void setOwnershipIdentityForCVPixelBuffer(CVPixelBufferRef pixelBuffer, const Pr
 
+    // FIXME: Remove workaround when rdar://158348697 is addressed
+    constexpr auto maxIOSurfaceWidth = 1 << 14;
+    if (width * height > maxIOSurfaceWidth * maxIOSurfaceWidth)
```
Should we use `safeMultiply` here?
Should we also message check `SharedVideoFrame::Buffer`?

mwyrzykowski replied:
yes good idea, I will make both changes

mwyrzykowski replied:
I made the first change and also increased the check to 1 << 15 since Apple10 hardware raises the limit to 32k x 32k

I'm looking to see where the message checks should be added

mwyrzykowski commented on Source/WebCore/platform/graphics/cv/CVUtilities.mm:
@youennf I made a bugzilla for the message check instead. There is some potential regression risk if done incorrectly. I'm concerned about making the change due to unfamiliarity with the code and potential to test all code paths correctly.

mwyrzykowski committed:
Limit CVPixelBuffer instances to reasonable sizes
https://bugs.webkit.org/show_bug.cgi?id=297470
rdar://158231386

Reviewed by Youenn Fablet and Jer Noble.

Limit CVPixelBuffer instances to reasonable sizes.

16k x 16k is the IOSurface maximum size and considerably
smaller than 8k video.

* Source/WebCore/platform/graphics/cv/CVUtilities.mm:
(WebCore::createBlackPixelBuffer):

Canonical link: https://commits.webkit.org/301112@main
